packages/calypso-e2e/src/lib/blocks/block-flows/ai-assistant.ts

```diff
@@ -0,0 +1,158 @@
+	 * configuring and validating the block.
+	 *
+	 * @param {ConfigurationData} configurationData Configuration data for the block.
+	 */
+	constructor( configurationData: ConfigurationData ) {
+		this.configurationData = configurationData;
+		this.generatedText = '';
+	}
+
+	blockSidebarName = 'AI Assistant (Experimental)';
+	blockEditorSelector = 'div[aria-label="Block: AI Assistant (Experimental)"]';
+
+	/**
+	 * Configure the block in the editor with the configuration data from the constructor
+	 *
+	 * @param {EditorContext} context The current context for the editor at the point of
+	 * test execution.
+	 */
+	async configure( context: EditorContext ): Promise< void > {
+		const editorCanvas = await context.editorPage.getEditorCanvas();
+		const block = editorCanvas.getByRole( 'document', {
+			name: `Block: ${ this.blockSidebarName }`,
+		} );
+
+		await this.enterInput( block );
+		await this.clickButtonOnBlock( block, 'Send' );
+		await this.waitForQuery( editorCanvas );
+
+		if ( this.configurationData.tone ) {
+			await this.clickToolbarButton( 'Change tone', this.configurationData.tone, context );
+			await this.waitForQuery( editorCanvas );
+		}
+
+		if ( this.configurationData.improve ) {
+			await this.clickToolbarButton( 'Improve', this.configurationData.improve, context );
+			await this.waitForQuery( editorCanvas );
+		}
+
+		await this.clickButtonOnBlock( block, 'Accept' );
+		await block.waitFor( { state: 'detached' } );
+		this.generatedText = await editorCanvas.locator( '.is-selected' ).innerText();
+	}
+
+	/**
+	 * Clicks on the toolbar button, then in the dropdown that appears, clicks an option.
+	 *
+	 * @param {string} buttonName Name of the button to click.
+	 * @param {string} optionName Name of the option in the dropdown to click.
+	 * @param {EditorContext} context The current context for the editor at the point of
+	 * test execution.
+	 */
+	private async clickToolbarButton(
+		buttonName: string,
+		optionName: string,
+		context: EditorContext
+	) {
+		// Block-based and non-block-based themes strike again.
+		// For block-based themes (eg. Twenty Twenty-Three) a given block's toolbar
+		// lives "outside" of the DOM tree returned by `getEditorCanvas`.
+		// For non-block-based themes (eg. Twenty Twenty-One) a block's toolbar
+		// lives in the same DOM tree as the locator returned by `getEditorCanvas`.
+		// However, the locator returned by `getEditorParent` has the toolbar inside
+		// the DOM tree in both cases.
+		const editorParent = await context.editorPage.getEditorParent();
+
+		const toolbar = editorParent.getByRole( 'toolbar', { name: 'Block tools' } );
+		await toolbar.waitFor();
+		// Yes, yuck, I know. But Playwright often tries to act on the buttons sooner
+		// than it should, so we need to wait for the bounding box of the element here
+		// to be stable before attempting to click on the toolbar.
+		const elementHandle = await toolbar.elementHandle();
```

Comment:
**Question (non-blocking):** Out of curiosity, can you explain what was going on here? I thought playwright auto-waited for `stable` on `click` actions, so I'm wondering if there's something else going on? 

https://playwright.dev/docs/actionability

Reply:
>  I thought playwright auto-waited for stable on click actions, so I'm wondering if there's something else going on?

Yes, Playwright does wait - and on the local machine, the spec doesn't need this little yucky part in order to pass _most of the time_. 

However, sometimes on local machine, and more frequently on CI, Playwright's auto-actionability checks simply race past the face that the Editor Toolbar isn't yet actually fully ready. 

As far as I can tell, the following happens in this scenario:

1. Playwright begins waiting for a button on the Editor Toolbar. AI Assistant block is generating and filling in text.
2. Once the AI Assistant block is done filling in text, the button begins to appear on the Editor Toolbar. 
3. Playwright resolves the waiting part, and begins actionability checks on the element. This, for some reason, passes when it should not, because the Editor Toolbar is still in the process of "expanding" in a sense, by rendering all the available action buttons.
4. Playwright thinks the target button is ready to click, so it performs the action. Meanwhile, the Editor Toolbar is not yet in fact ready, but the action gets performed on it anyway. 
5. The Editor Toolbar ends up "swallowing" the action, and the expected Popover doesn't show.
6. Playwright fails to locate the option in the Popover because it isn't there.

Reply:
Case in point, I was able to get this portion to fail relatively consistently on local machine in Jetpack AT environment, and it stopped as soon as I put the `waitForElementState( stable )` call.

To me, it seems the Playwright's actionability waits only apply to the immediate locator you're trying to action upon. For example, in this case, the actionability checks determine the "Change tone" button is present and it's ready to interact with, but in reality the parent component "Editor Toolbar" is not yet stable.

Reply:
In the most recent revision I removed the use of ElementHandle and it seems to still do fine, so let's stick with that for now.

Reply:
Ah sure, thank you so much for explaining all that! That makes sense, and is a bummer to have to work with! 😕 There could also be a React race condition here too where the element is in the DOM, but the handlers attached to it aren't included in the first render pass or something like that. Thanks for explaining! If it's running stable, I think we're good to go 👍  😄